File: real_benchmarks/data/real_c_rust_pairs/aliasing_violation.c
/*
 * Strict aliasing violation — inspired by real-world type punning bugs.
 * CVE-2008-1447 (DNS cache poisoning exploited via pointer aliasing in bind).
 *
 * C11 §6.5/7 (strict aliasing rule): an object shall have its stored value
 * accessed only by an lvalue of a compatible type. Accessing int through
 * float pointer (or vice versa) is UB.
 *
 * Rust: raw pointer casts require unsafe, and Miri catches violations.
 */
#include <stdint.h>
#include <string.h>

/* Classic strict aliasing violation: int → float through pointer cast. */
float int_to_float_aliased(int32_t x) {
    float *fp = (float *)&x;  /* violates strict aliasing */
    return *fp;  /* UB: reading int object through float lvalue */
}

/* Reverse: float → int through pointer cast. */
int32_t float_to_int_aliased(float x) {
    int32_t *ip = (int32_t *)&x;  /* violates strict aliasing */
    return *ip;  /* UB: same issue */
}

/* Safe version using memcpy (not UB — type-punning via memcpy is legal). */
float int_to_float_safe(int32_t x) {
    float result;
    memcpy(&result, &x, sizeof(result));  /* legal: memcpy is special */
    return result;
}

/* TBAA (Type-Based Alias Analysis) exploit: compiler assumes no aliasing. */
void swap_values(int *restrict a, float *restrict b) {
    /* Compiler assumes a and b don't alias, but if called with overlapping
       memory, the optimization may produce wrong results. */
    int tmp_a = *a;
    float tmp_b = *b;
    *(float *)a = tmp_b;  /* UB: writes float to int object */
    *(int *)b = tmp_a;    /* UB: writes int to float object */
}

/* char* is allowed to alias anything in C (the char exception). */
int read_through_char(int *p) {
    char *cp = (char *)p;  /* legal: char can alias any type */
    int sum = 0;
    for (int i = 0; i < (int)sizeof(int); i++) {
        sum += cp[i];
    }
    return sum;
}

File: real_benchmarks/data/real_c_rust_pairs/aliasing_violation.rs
/*
 * Rust translation of aliasing violation patterns.
 * Rust's type system and borrow checker prevent aliasing violations.
 * Raw pointer casts require unsafe and are checked by Miri.
 */

/// Int to float reinterpretation: Rust uses f32::from_bits (safe).
pub fn int_to_float_aliased(x: i32) -> f32 {
    f32::from_bits(x as u32)  // safe, defined behavior, no aliasing issue
}

/// Float to int reinterpretation: Rust uses f32::to_bits (safe).
pub fn float_to_int_aliased(x: f32) -> i32 {
    x.to_bits() as i32  // safe, no pointer aliasing
}

/// Safe version: identical to aliased version in Rust — both are safe.
pub fn int_to_float_safe(x: i32) -> f32 {
    f32::from_bits(x as u32)
}

/// Swap: Rust's borrow checker prevents mutable aliasing at compile time.
/// &mut a and &mut b are guaranteed not to overlap.
pub fn swap_values(a: &mut i32, b: &mut f32) {
    let tmp_a = *a;
    let tmp_b = *b;
    *b = tmp_a as f32;
    *a = tmp_b as i32;
}

/// Byte-level access: Rust uses to_ne_bytes() for safe byte access.
pub fn read_through_char(p: &i32) -> i32 {
    let bytes = p.to_ne_bytes();
    let mut sum: i32 = 0;
    for &b in bytes.iter() {
        sum = sum.wrapping_add(b as i32);
    }
    sum
}
